sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,347 @@
+      );
+      WholeFileIO.ReadAll readAll = readAll();
+      return input.apply(Create.ofProvider(getFilePattern(), StringUtf8Coder.of()))
+          .apply("Via ReadAll", readAll);
+    }
+  }
+
+  /** Implementation of {@link #readAll()}. */
+  @AutoValue
+  public abstract static class ReadAll extends PTransform<PCollection<String>,
+                                                                  PCollection<KV<String, byte[]>>> {
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract ReadAll build();
+    }
+
+    @Override
+    public PCollection<KV<String, byte[]>> expand(PCollection<String> input) {
+      return input.apply(Match.filepatterns())
+          .apply(ParDo.of(new ReadWholeFileFn()));
+    }
+
+    /**
+     * A {@link DoFn} that takes in {@link MatchResult.Metadata} from which it acquires a
+     * {@link ResourceId} that it reads into a {@code byte[]} paired in a {@link KV} with the
+     * filename as a {@link String}.
+     */
+    public static class ReadWholeFileFn extends DoFn<MatchResult.Metadata, KV<String, byte[]>> {
+      @ProcessElement
+      public void processElement(ProcessContext c) throws IOException {
+        MatchResult.Metadata metadata = c.element();
+        ResourceId resourceId = metadata.resourceId();
+
+        try (InputStream inStream = Channels.newInputStream(FileSystems.open(resourceId))) {
+          c.output(KV.of(resourceId.getFilename(), StreamUtils.getBytes(inStream)));
+        }
+      }
+    }
+  }
+
+  /** Implementation of {@link #write()}. */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<KV<String, byte[]>>, PDone> {
+    private static final Logger LOG = LoggerFactory.getLogger(Write.class);
+
+    @Nullable abstract ValueProvider<ResourceId> getOutputDir();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setOutputDir(ValueProvider<ResourceId> outputDir);
+
+      abstract Write build();
+    }
+
+    public Write to(String outputDir) {
+      return to(FileSystems.matchNewResource(outputDir, true));
+    }
+
+    public Write to(ResourceId outputDir) {
+      return toResource(ValueProvider.StaticValueProvider.of(outputDir));
+    }
+
+    public Write toResource(ValueProvider<ResourceId> outputDir) {
+      return toBuilder().setOutputDir(outputDir).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<String, byte[]>> input) {
+      checkNotNull(
+          getOutputDir(),
+          "Need to set the output directory of a WholeFileIO.Write transform."
+      );
+
+      input.apply(
+          ParDo.of(
+              new DoFn<KV<String, byte[]>, Void>() {
+                ValueProvider<ResourceId> tmpDir;
+
+                @Setup
+                public void setup() {
+                  tmpDir = ValueProvider.NestedValueProvider.of(
```

Comment:
Hm this seems unnecessary - both having a `ValueProvider` variable, and initializing it in `setup`.

Instead, in `processElement`, you can do simply:

ResourceId outputDir = getOutputDir().get();
ResourceId tmpDir = buildTmpDir(outputDir);  // create a static function that does what TemporaryDirectoryBuilder.apply() does, and remove TemporaryDirectoryBuilder